fix(ci): prevent Playwright snapshot loop

webjunkie opened the pull request:
# Fix Playwright E2E Snapshot Loop with Flapping Detection

## Problem

Some PRs get stuck in infinite snapshot loops where the bot continuously updates screenshots. Example: PR #24707 shows repeated commits 10 minutes apart, wasting CI resources and preventing PRs from stabilizing.

## Solution

**Detect flapping screenshots and fail CI** to require human intervention rather than allowing infinite loops.

### How It Works

1. **First bot commit**: Screenshot differs → Bot commits update → Workflow re-runs
2. **Second bot commit** (same screenshot): Flapping detected → CI fails → Merge blocked → Human must fix

### Changes Made

1. **Enhanced flap detection** - Counts bot changes per screenshot since last human commit
2. **Fail on flapping** - Workflow exits with error when 2+ bot changes detected
3. **PR comment** - Explains flapping and how to fix it
4. **Block commits** - Don't commit new snapshots when flapping detected

## Benefits

- ✅ Breaks infinite loops after 2 bot commits
- ✅ Blocks automerge when tests unstable
- ✅ Forces fixing actual flakiness (not masking symptoms)
- ✅ Snapshots still verified (re-runs happen, just limited)

## Trade-offs

- One extra CI run per legitimate snapshot update (acceptable cost)
- Developers must fix flaky tests when detected (good forcing function)

<details>
<summary>Scenario Analysis</summary>

### Legitimate One-Time Update (Happy Path)

1. Developer pushes code
2. Screenshot differs → Bot commits (1st change)
3. Workflow re-runs → Screenshot stable → Passes ✅

### Flapping Screenshot (Circuit Breaker)

1. Screenshot differs → Bot commits v1 (1st change)
2. Re-run → Differs again → Flapping detected (2nd change)
3. CI fails ❌ → Merge blocked → Human intervenes

### Broken UI with Automerge

1. Broken UI pushed → Many screenshots fail
2. Bot commits all (1st change each)
3. Re-run → If truly broken, screenshots still differ
4. Flapping detected → CI fails ❌ → **Automerge blocked** ✅

### Parallel PRs - Same Screenshot

1. PR A and PR B both modify button.png independently
2. PR A merges first
3. PR B rebases → Git conflict (binary file)
4. Developer resolves → Push triggers E2E
5. Bot commits merged version → Re-run verifies → Passes ✅

### Stable But Wrong Screenshot

If broken UI produces a *stable* wrong screenshot:
- Screenshot IS stable (not flapping) → Passes
- This is a code review problem, not CI problem
- Human reviewer should catch visual regression
- Not a regression from current behavior

</details>

<details>
<summary>Implementation Details</summary>

### Flap Detection Logic

```bash
for screenshot in $(git diff --name-only HEAD playwright/__snapshots__); do
  bot_changes=$(git log --oneline ${last_human_commit}..HEAD --author="github-actions" -- $screenshot | wc -l)
  if [ "$bot_changes" -ge 2 ]; then
    FLAPPING=true
  fi
done
```

### Workflow Steps

1. **Detect flapping** - Count bot changes per screenshot
2. **Comment on PR** - If flapping detected, explain issue
3. **Fail workflow** - Exit 1 if flapping
4. **Commit snapshots** - Only if NOT flapping

### For Developers

**If bot commits flapping screenshots:**

1. Investigate test flakiness (timing, animations, etc.)
2. Fix the underlying issue
3. Verify locally (run tests multiple times)
4. Push your fix (resets flap counter)

**Manual trigger if needed:**
- Actions tab → E2E CI Playwright → Run workflow

</details>

## Testing

This PR includes a small test change to trigger snapshot updates and verify the flapping detection works as expected.

webjunkie committed:
fix(ci): prevent Playwright snapshot loop by excluding snapshots from workflow trigger

Adds path filter exclusion for playwright/__snapshots__/** to prevent
infinite loops where bot commits trigger new workflow runs.

Changes:
- Add workflow_dispatch trigger for manual runs
- Exclude playwright/__snapshots__/** from path filter
- Add PR comment explaining no re-run will occur after snapshot updates

Problem:
Screenshot updates by the bot were triggering new E2E runs, which would
sometimes produce slightly different screenshots due to timing or
rendering variations, causing an infinite loop.

Solution:
Snapshots are now excluded from triggering workflows. They're still
verified naturally when developers push changes, rebase, or when
subsequent PRs run E2E tests. This provides more meaningful verification
(different timing/environment) than immediate re-runs.

webjunkie commented on .github/workflows/ci-e2e-playwright.yml:
I had it evaluate and document various scenarios 😆 

I thought breaking the loop would put and end to it, but you are right, auto-merging is a valid point. Let me think about this.

webjunkie committed:
fix(ci): prevent snapshot loops with flapping detection

Replace path filter exclusion with circuit breaker that detects flapping
screenshots and fails CI to require human intervention. This prevents both
infinite loops and broken UI from merging via automerge.

How it works:
- First bot commit: Screenshot differs → Bot commits → Re-runs
- Second bot commit (same screenshot): Flapping detected → CI fails
- Human must fix underlying test flakiness before merge can proceed

Changes:
- Enhanced flap detection to count bot changes per screenshot
- Added workflow failure step when flapping detected
- Added PR comment explaining flapping and how to fix
- Conditioned snapshot commits on not flapping
- Removed old PR comment about no re-runs (no longer accurate)
- Reverted path filter exclusion (need re-runs to detect flapping)

webjunkie committed:
test(e2e): add visual regression test to trigger snapshot update

This test will trigger a snapshot update when run in CI, allowing us to
observe and verify the flapping detection circuit breaker works correctly.

The test should:
1. First run: Create new snapshot (bot commits)
2. Second run: Verify snapshot matches (passes)
3. If flapping: Detection triggers, CI fails

This is a temporary test for verification purposes.

webjunkie committed:
test(e2e): make flapping test actually flap with timestamp

The test now injects a timestamp div into the page that changes on every
run, guaranteed to cause screenshot flapping. This will trigger the flapping
detection circuit breaker on the second bot commit.

Expected behavior in CI:
1. First run: Creates snapshot with timestamp A, bot commits
2. Second run: Timestamp B differs, flapping detected (2 bot changes)
3. CI fails with flapping error, merge blocked
4. Human must remove this test or fix it before merge

github-actions[bot] committed:
Update E2E screenshots for Playwright

github-actions[bot] committed:
Update E2E screenshots for Playwright

webjunkie committed:
fix(ci): improve flap detection to properly find last human commit

The perl-regexp pattern wasn't working, causing last_human_commit to be empty.
Now using simpler grep -v approach that filters out github-actions author.

Also added safety check for empty last_human_commit to prevent false positives.

github-actions[bot] committed:
Update E2E screenshots for Playwright

webjunkie committed:
fix(ci): fetch enough git history for flap detection

Checkout was shallow (depth=1), causing git log to fail finding human commits.
Now fetching 50 commits which covers most PRs while staying fast.

webjunkie committed:
fix(ci): add proper error handling to flap detection

- set +e to prevent script exit on grep failures
- || true on grep command to handle no-match case gracefully
- Better error messaging for debugging

The script was failing with exit code 1 when grep -v found no matches.

github-actions[bot] committed:
Update E2E screenshots for Playwright

github-actions[bot] committed:
Update E2E screenshots for Playwright

webjunkie committed:
fix(ci): post flapping comment immediately when detected

Remove failure() condition from PR comment step. The comment should post
as soon as flapping is detected, not after workflow fails. This ensures
developers always get the explanatory message.

Thanks to reviewer for catching this!

github-actions[bot] committed:
Update E2E screenshots for Playwright

github-actions[bot] committed:
Update E2E screenshots for Playwright

webjunkie committed:
fix(ci): fix syntax error in PR comment script

Nested template literals with backtick escaping was causing:
'SyntaxError: Invalid or unexpected token'

Extracted path list formatting to separate variable to avoid escaping issues.

github-actions[bot] committed:
Update E2E screenshots for Playwright

github-actions[bot] committed:
Update E2E screenshots for Playwright

pauldambra commented on playwright/e2e/system-status.spec.ts:
```diff
@@ -6,4 +6,25 @@ test.describe('System Status', () => {
         await expect(page.locator('table')).toHaveText(/Events in ClickHouse/)
     })
+
+    test('System Status visual regression with timestamp', async ({ page }) => {
```
i guess we need to unflap this test 😊

pauldambra reviewed (approved):
this seems a way safer approach

worst case someone can ask for a force merge if they _can't_ spend the time fixing
and that's true today _anyway_ since we're having to wait for the flapchain to settle

webjunkie committed:
test(e2e): remove flapping test after successful verification

The flapping detection circuit breaker has been tested and works:
- Detected 2 bot commits to the same screenshot
- Posted PR comment explaining the issue
- Failed workflow to block merge

Test no longer needed.
